util/fetch/src/client.rs

```diff
@@ -14,347 +14,554 @@
+	pub fn abort(&self) {
+		self.abort.store(true, Ordering::SeqCst)
+	}
 
-	/// Fetch URL and get a future for the result.
-	fn fetch(&self, url: &str) -> Self::Result {
-		self.fetch_with_abort(url, Default::default())
+	/// Set the maximum reponse body size.
+	pub fn with_max_size(self, n: usize) -> Abort {
+		Abort { size: n, .. self }
 	}
 
-	/// Fetch URL and get the result synchronously.
-	fn fetch_sync(&self, url: &str) -> Result<Response, Error> {
-		self.fetch(url).wait()
+	/// Set the maximum duration (including redirects).
+	pub fn with_max_duration(self, d: Duration) -> Abort {
+		Abort { time: d, .. self }
 	}
 
-	/// Closes this client
-	fn close(self) where Self: Sized {}
+	/// Set the maximum number of redirects to follow.
+	pub fn with_max_redirects(self, n: usize) -> Abort {
+		Abort { redir: n, .. self }
+	}
 }
 
-const CLIENT_TIMEOUT_SECONDS: u64 = 5;
+/// Types which retrieve content from some URL.
+pub trait Fetch: Clone + Send + Sync + 'static {
+	/// The result future.
+	type Result: Future<Item=Response, Error=Error> + Send + 'static;
 
-/// Fetch client
+	/// Get content from some URL.
+	fn fetch(&self, url: &str, abort: Abort) -> Self::Result;
+}
+
+type TxResponse = oneshot::Sender<Result<Response, Error>>;
+type StartupCond = Arc<(Mutex<Result<(), io::Error>>, Condvar)>;
+type ChanItem = Option<(Url, Abort, TxResponse)>;
+
+/// An implementation of `Fetch` using a `hyper` client.
+// Due to the `Send` bound of `Fetch` we spawn a background thread for
+// actual request/response processing as `hyper::Client` itself does
+// not implement `Send` currently.
+#[derive(Debug)]
 pub struct Client {
-	client: RwLock<(time::Instant, Arc<reqwest::Client>)>,
-	pool: CpuPool,
-	limit: Option<usize>,
+	core: mpsc::Sender<ChanItem>,
+	refs: Arc<AtomicUsize>,
 }
 
+// When cloning a client we increment the internal reference counter.
 impl Clone for Client {
-	fn clone(&self) -> Self {
-		let (ref time, ref client) = *self.client.read();
+	fn clone(&self) -> Client {
+		self.refs.fetch_add(1, Ordering::SeqCst);
 		Client {
-			client: RwLock::new((time.clone(), client.clone())),
-			pool: self.pool.clone(),
-			limit: self.limit.clone(),
+			core: self.core.clone(),
+			refs: self.refs.clone(),
 		}
 	}
 }
 
-impl Client {
-	fn new_client() -> Result<Arc<reqwest::Client>, Error> {
-		let mut client = reqwest::ClientBuilder::new();
-		client.redirect(reqwest::RedirectPolicy::limited(5));
-		Ok(Arc::new(client.build()?))
+// When dropping a client, we decrement the reference counter.
+// Once it reaches 0 we terminate the background thread.
+impl Drop for Client {
+	fn drop(&mut self) {
+		if self.refs.fetch_sub(1, Ordering::SeqCst) == 1 {
+			// ignore send error as it means the background thread is gone already
+			let _ = self.core.clone().send(None).wait();
+		}
 	}
+}
+
+impl Client {
+	/// Create a new fetch client.
+	pub fn new() -> Result<Self, Error> {
+		let startup_done = Arc::new((Mutex::new(Ok(())), Condvar::new()));
```

Comment:
I suppose `std::sync::mpsc::channel/sync_channel()` would be a bit simpler here, and also supports timeouts.